File: frontend/src/tipos/combinadas.ts
/**
 * combinadas.ts — Tipos para apuestas combinadas
 */

import { NivelConfianza } from './analisis';
import { ResultadoApuesta } from './bitacora';

export type ResultadoCombinada = 'PENDIENTE' | 'GANADA' | 'PERDIDA' | 'PUSH';

export interface SeleccionCombinadaInput {
  partido_id?: string | null;
  equipo_local: string;
  equipo_visitante: string;
  fecha_partido?: string | null;
  mercado: 'Q1' | 'Q2' | 'Q3' | 'Q4' | 'COMPLETO';
  lado: 'OVER' | 'UNDER';
  linea: number;
  cuota: number;
  cuota_over?: number | null;
  cuota_under?: number | null;
  probabilidad_sistema: number;
  prediccion_media?: number | null;
  prediccion_desviacion?: number | null;
  confianza_sistema?: NivelConfianza | null;
  valor_esperado_individual?: number | null;
  razones?: Array<Record<string, unknown>> | null;
}

export interface SeleccionCombinada extends SeleccionCombinadaInput {
  id: string;
  combinada_id: string;
  grupo_correlacion?: number | null;
  factor_ajuste_correlacion?: number | null;
  resultado?: ResultadoApuesta | null;
  puntos_reales?: number | null;
  fecha_resolucion?: string | null;
  orden: number;
}

export interface Combinada {
  id: string;
  usuario_id: string;
  stake: number;
  cuota_total: number;
  n_selecciones: number;
  probabilidad_independiente: number;
  factor_correlacion: number;
  probabilidad_ajustada: number;
  cuota_justa_sistema?: number | null;
  edge_combinada?: number | null;
  valor_esperado?: number | null;
  confianza_sistema?: NivelConfianza | null;
  tiene_mismo_partido: boolean;
  n_partidos_unicos: number;
  advertencias?: string[] | null;
  correlaciones_detectadas?: Array<Record<string, unknown>> | null;
  resultado: ResultadoCombinada;
  ganancia: number;
  selecciones_ganadas: number;
  selecciones_perdidas: number;
  selecciones_push: number;
  selecciones_pendientes: number;
  fecha_resolucion?: string | null;
  notas?: string | null;
  etiquetas?: string[] | null;
  creado_en?: string | null;
  actualizado_en?: string | null;
  selecciones?: SeleccionCombinada[];
}

export interface PeticionCrearCombinada {
  stake: number;
  cuota_total: number;
  selecciones: SeleccionCombinadaInput[];
  notas?: string | null;
  etiquetas?: string[] | null;
}

export interface RespuestaCombinada {
  exito: boolean;
  combinada: Combinada;
}

export interface RespuestaListaCombinadas {
  exito: boolean;
  total: number;
  pagina: number;
  total_paginas: number;
  combinadas: Combinada[];
}

export type TipoApuesta = 'SIMPLE' | 'COMBINADA';

export interface RegistroBitacoraUnificada {
  id: string;
  tipo_apuesta: TipoApuesta;
  resultado: ResultadoApuesta | ResultadoCombinada;
  stake: number;
  ganancia: number;
  cuota?: number | null;
  cuota_total?: number | null;
  n_selecciones?: number | null;
  selecciones_ganadas?: number | null;
  selecciones_perdidas?: number | null;
  selecciones_push?: number | null;
  selecciones_pendientes?: number | null;
  tiene_mismo_partido?: boolean | null;
  factor_correlacion?: number | null;
  advertencias?: string[] | null;
  equipo_local?: string | null;
  equipo_visitante?: string | null;
  fecha_partido?: string | null;
  mercado?: string | null;
  lado?: string | null;
  linea?: number | null;
  probabilidad_sistema?: number | null;
  confianza_sistema?: NivelConfianza | null;
  valor_esperado?: number | null;
  creado_en?: string | null;
  actualizado_en?: string | null;
  selecciones?: SeleccionCombinada[] | null;
}

export interface RespuestaBitacoraUnificada {
  exito: boolean;
  total: number;
  pagina: number;
  total_paginas: number;
  registros: RegistroBitacoraUnificada[];
}

File: backend/esquemas/combinadas.py
# -*- coding: utf-8 -*-
"""
esquemas/combinadas.py — Esquemas Pydantic para combinadas.
"""

from __future__ import annotations

from datetime import date
from typing import List, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator


class SeleccionCombinadaBase(BaseModel):
    """Datos base de una selección dentro de una combinada."""

    partido_id: Optional[UUID] = Field(None, description="ID del partido")
    equipo_local: str = Field(..., min_length=1)
    equipo_visitante: str = Field(..., min_length=1)
    fecha_partido: Optional[date] = Field(None)
    mercado: Literal["Q1", "Q2", "Q3", "Q4", "COMPLETO"]
    lado: Literal["OVER", "UNDER"]
    linea: float = Field(..., gt=0)
    cuota: float = Field(..., gt=1.0)
    cuota_over: Optional[float] = Field(None, gt=1.0)
    cuota_under: Optional[float] = Field(None, gt=1.0)
    probabilidad_sistema: float = Field(..., ge=0.0, le=1.0)
    prediccion_media: Optional[float] = None
    prediccion_desviacion: Optional[float] = None
    confianza_sistema: Optional[Literal["ALTA", "MEDIA", "BAJA"]] = None
    valor_esperado_individual: Optional[float] = None
    razones: Optional[list] = None

    @field_validator("equipo_local", "equipo_visitante")
    @classmethod
    def validar_equipo(cls, valor: str) -> str:
        if not valor.strip():
            raise ValueError("El nombre del equipo no puede estar vacío.")
        return valor


class SeleccionCombinadaCrear(SeleccionCombinadaBase):
    """Entrada para crear una selección de combinada."""


class SeleccionCombinada(SeleccionCombinadaBase):
    """Salida de una selección de combinada."""

    id: UUID
    combinada_id: UUID
    grupo_correlacion: Optional[int] = None
    factor_ajuste_correlacion: Optional[float] = None
    resultado: Optional[Literal["PENDIENTE", "GANADA", "PERDIDA", "PUSH", "ANULADA"]] = None
    puntos_reales: Optional[float] = None
    fecha_resolucion: Optional[str] = None
    orden: int


class PeticionCrearCombinada(BaseModel):
    """Solicitud para crear una combinada."""

    stake: float = Field(..., gt=0)
    cuota_total: float = Field(..., gt=1.0)
    selecciones: List[SeleccionCombinadaCrear]
    notas: Optional[str] = None
    etiquetas: Optional[List[str]] = None


class Combinada(BaseModel):
    """Modelo completo de combinada."""

    id: UUID
    usuario_id: UUID
    stake: float
    cuota_total: float
    n_selecciones: int
    probabilidad_independiente: float
    factor_correlacion: float
    probabilidad_ajustada: float
    cuota_justa_sistema: Optional[float] = None
    edge_combinada: Optional[float] = None
    valor_esperado: Optional[float] = None
    confianza_sistema: Optional[str] = None
    tiene_mismo_partido: bool
    n_partidos_unicos: int
    advertencias: Optional[List[str]] = None
    correlaciones_detectadas: Optional[List[dict]] = None
    resultado: Literal["PENDIENTE", "GANADA", "PERDIDA", "PUSH", "ANULADA"]
    ganancia: float
    selecciones_ganadas: int
    selecciones_perdidas: int
    selecciones_push: int
    selecciones_pendientes: int
    fecha_resolucion: Optional[str] = None
    notas: Optional[str] = None
    etiquetas: Optional[List[str]] = None
    creado_en: Optional[str] = None
    actualizado_en: Optional[str] = None
    selecciones: Optional[List[SeleccionCombinada]] = None


class RespuestaCombinada(BaseModel):
    """Respuesta con una combinada."""

    exito: bool
    combinada: Combinada


class RespuestaListaCombinadas(BaseModel):
    """Respuesta al listar combinadas."""

    exito: bool
    total: int
    pagina: int
    total_paginas: int
    combinadas: List[Combinada]


class PeticionActualizarResultadoCombinada(BaseModel):
    """Solicitud para actualizar el resultado de una combinada."""

    resultado: Literal["GANADA", "PERDIDA", "PUSH", "ANULADA"] = Field(
        ..., description="Resultado de la combinada"
    )
